packages/remix-react/transition.ts

```diff
@@ -626,17 +626,21 @@ export function createTransitionManager(init: TransitionManagerInit) {
       if (updates.transition === IDLE_TRANSITION) {
         pendingNavigationController = undefined;
 
-        // Preserve deferred data for routes that were not cancelled, or were
-        // cancelled and returned a subsequent deferred() in actionReload
+        // Preserve deferred data for active routes that were not cancelled, or
+        // were cancelled and are again pending (i.e., returned a subsequent
+        // deferred() in actionReload)
         overrides.deferredLoaderData = [
           ...Object.entries(state.deferredLoaderData),
           ...Object.entries(updates.deferredLoaderData || {}),
         ]
-          .filter(
-            ([routeId]) =>
-              !cancelledDeferredRouteIds.has(routeId) ||
-              pendingNavigationDeferredControllers.has(routeId)
-          )
+          .filter(([routeId]) => {
+            let isActive = (updates.matches || state.matches).some(
+              (m) => m.route.id === routeId
+            );
```

Comment:
Added this check here to make sure that we only preserve `deferredLoaderData` for actively matched routes.  We were sneaking by ok previously because we were not removing from `pendingNavigationDeferredControllers` upon resolution, which meant that on subsequent abort's we would put it into `cancelledDeferredRouteIds` (even though it wasn't actually cancelled) and that would prevent us from carrying forward the `deferredLoaderData`.  

For example:
 - navigate `/foo` which returns a `deferred`
 - foo `deferred` completely resolved, but remains in `pendingNavigationDeferredControllers`
 - navigate `/bar`
 - foo is "cancelled" because it no longer matches
 - when /bar completes we do not retain foo because it was cancelled

Instead, we should not be marking foo as cancelled and should instead mark it as "not pending".  Then we need to identify that it's not active here and drop it's `deferredLoaderData`